src/libsystemd/sd-journal/journal-file.c

```diff
@@ -590,7 +590,8 @@ static unsigned type_to_context(ObjectType type) {
         return type > OBJECT_UNUSED && type < _OBJECT_TYPE_MAX ? type : 0;
 }
 
-static int journal_file_move_to(
+/* XXX keep this inline, this is on a very hot path for things like journalctl matches */
+static inline int journal_file_move_to(
```

Comment:
as a general rule: we don't use "static inline" inside of .c files. The compiler is free to ignore the "inline" anyway, and in fact gcc doesn't really care much: it's able to detect on its own better when it's worth inlining something.

"static inline" has uses in .h files, since it allows defining inline-worthy functions in header files with full body. But outside of .h files it simply shouldn't be used, just using "static" is entirely enough, and gcc will likely inline single-use, static functions anyway.

Reply:
Are you compiling with -O0? If you do, gcc takes things literally, to make debugging easy. But that's not how production software is built...

Reply:
I'm just using `meson setup --buildtype=release builddir; meson compile -C builddir`, no explicit manipulation of CFLAGS is going on.  Aren't the release builds optimized? Am I holding it wrong?

Reply:
It appears when simply using `meson configure --buildtype=release` the optimization level is `-O1`.

And glancing at Arch and Debian source packages, if those qualify as "how production software is built", simply setting the build mode seems to be the norm, unless I'm missing something.

I just want optimization levels representative of what people are likely to be experiencing in common distros out there, the whole point is to characterize and improve if needed the performance for the common case.

Reply:
Just rebuilt v247, v248, and main+this PR with more explicit meson setup options `-Dmode=release -Ddebug=false -Doptimization=3 -D b_lto=true -D b_pie=true`, and here's the best times of four runs each:

```
$ time LD_LIBRARY_PATH=build-v247-release-O3-lto-pie/ build-v247-release-O3-lto-pie/journalctl --directory rel-builds/test-journals/ -u postfix@- > /dev/null

real    0m10.114s
user    0m9.978s
sys     0m0.120s


$ time LD_LIBRARY_PATH=build-v248-release-O3-lto-pie/ build-v248-release-O3-lto-pie/journalctl --directory rel-builds/test-journals/ -u postfix@- > /dev/null

real    0m10.708s
user    0m10.573s
sys     0m0.127s

~5.5% slowdown v247->v248

$ time LD_LIBRARY_PATH=build-inlines-PR-release-O3-lto-pie/ build-inlines-PR-release-O3-lto-pie/journalctl --directory rel-builds/test-journals/ -u postfix@- > /dev/null

real    0m9.848s
user    0m9.721s
sys     0m0.114s

8% speedup v248->main+this PR
```

Still significant differences there, smaller than the -O1 from --buildtype=release, but multiple whole percents.

It's unclear to me which distros are explicitly setting -Doptimization=3 BTW.  Maybe they have hidden stuff happening in meson setup wrappers, but at least in the Arch and Debian source packages what's visible is just a mode=release.